### PR TITLE
Added support in Mavlink Ethernet channel parameters

### PR DESCRIPTION
**Pull request by garfieldG**
Mavlink Ethernet channel settings such as UDP port, remote port and broadcast mode now can be changed dynamically via parameters.

**Describe problem solved by this pull request**
-The start of the Mavlink Ethernet channel is static and not dynamic like the start of the serial channel.
-It allows us to dynamically choose Mavlink Ethernet channels settings via parameters.
-It allows us to use Broadcast on one Mavlink channel and Multicast on another.

**Describe your solution**
I added Ethernet related parameters to Mavlink.
Just like the serial parameters we had, the new parameters are times the number of Mavlink instances and are generated.
The new parameters will be generated only if the board supports Ethernet (by adding ETHERNET to its cmake).
Also, to choose Ethernet a new value was added to MAV_{I}_CONFIG and will be available only if the board supports Ethernet (otherwise the new value won't be generated as an option).

**Additional context**
I tested it on a costume board and it works fine.
Previously I [suggested](https://github.com/PX4/Firmware/pull/10124) adding parameters such as IP address, Mac address.
In my code, I also added parameters to choose remote multicast address and another line in the generated Mavlink command that supports multicast, however, the added line only works if the multicast address is known from the parameters. I will gladly add it if you want.

### CODE REVIEW

**Comment by dagar** [comment]
Slightly lazy question, how does this extend multiple instances? For example on SITL (or a linux board) where we have multiple instances on different ports. Does specifying the board has ETHERNET give you one additional (parameter configurable and autostarted) Mavlink instance?

**Comment by dagar** [comment]
@dinomani this might be of interest for v5x.

**Comment by garfieldG** [comment]
@dagar Hi, I'm not sure that I understand your question, but I'll try to answer.
It doesn't give any extra instances, you still have the same number of possible Mavlink instances, but you are now able to configure any of these Mavlink instances to use ethernet instead of serial via parameters.
When you specify your board has ETHERNET :
- ethernet is added as a value that you can select :
image
you can choose ethernet on one instance or more. 
not like the other values, ethernet can be chosen multiple times.
 -new parameters are added :
image
image
image

note that the extra parameters and value are to all instances and not just to instance 0.

one more thing, this change is not only for Mavlink, if in the future there will be any parameters that you would like to generate only for boards that support ethernet then you can always do that by using the ethernet keyword inside the yaml file.

**Comment by garfieldG** [comment]
@dagar @dinomani 
Hi, 
Is there anything new regarding this PR? Would you like me to add or change something?

**Comment by BazookaJoe1900** [comment]
@dagar can you take a look on that?

**Comment by dagar** [comment]
Sorry, I don't recall where we got to with this one.

Having a configurable ports per mavlink instance feels wrong, but we can live with it. What if I wanted to have more than 1 mavlink instance (eg SITL)?

**Comment by garfieldG** [comment]
> Having configurable ports per mavlink instance feels wrong
I'm not sure I agree with you about that. I did it because we encountered a problem in which we needed more than one mavlink instance and each one of the instances needed to be working with a different port. It needed to be configurable because our need changed from time to time so we wanted it to be parameters configurable with no need to change something in the code each time just for a different port.
> What if I wanted to have more than 1 mavlink instance
No problem here. Just like it was before you have MAV_{i}_CONFIG. You add mavlink instances just like you did it before, the only change is that now you can select that your mavlink instances will work with ethernet instead of serial. you can have more than one ethernet instance, because unlike the regular serial options you can select in MAV_{i}_CONFIG the ethernet option can be selected multiple times, if you chose it once you can choose it again as many times that you want (and as many max instances you have of course). Also, the ethernet option will only appear as a possible option only if your board supports ethernet (PX4_ETHERNET).

**Comment by BazookaJoe1900** [comment]
@dagar ping....

File: src/modules/mavlink/module.yaml [hunk 1]
```diff
@@ -95,3 +104,45 @@ parameters:
...
+                    If ethernet enabled and selected as configuration for MAVLink instance ${i}, 
+                    selected udp port will be set and used in MAVLink instance ${i}.
+
+            type: int32
+            reboot_required: true
+            num_instances: *max_num_config_instances
+            default: [14556, 0, 0]
+            ethernet:
+
+        MAV_${i}_REMOT_PORT:
```
Comment (hamishwillee) [inline src/modules/mavlink/module.yaml:121]: REMOT or REMOTE?

**Comment by hamishwillee** [comment]
If this goes in it may need documentation. Probably and update to https://docs.px4.io/master/en/peripherals/mavlink_peripherals.html

File: src/modules/mavlink/module.yaml (hunk 1 above)
Comment (garfieldG) [reply src/modules/mavlink/module.yaml:121]: @hamishwillee REMOT... REMOTE will bring us to 17 chars and 16 is the limit

Comment (hamishwillee) [reply src/modules/mavlink/module.yaml:121]: Thanks @garfieldG I guess I should have counted. FWIW I probably would have done REMOTE_PRT, although I guess a matter of taste

Comment (BazookaJoe1900) [reply src/modules/mavlink/module.yaml:121]: I agree with @hamishwillee REMOTE_PRT looks better

Comment (garfieldG) [reply src/modules/mavlink/module.yaml]: @hamishwillee Changed it. Also changed UDP_PORT to UDP_PRT to be similar

Comment (hamishwillee) [reply src/modules/mavlink/module.yaml]: Thanks @garfieldG .I wouldn't have suggested you do that (even for consistency) but it is completely fine :-).

**Comment by LorenzMeier** [comment]
Could you rebase? Sorry that this didn't get pushed forward earlier.

**Comment by garfieldG** [comment]
@LorenzMeier Thanks for the update. Just rebased it now.

**Comment by LorenzMeier** [comment]
Thanks!

**Comment by garfieldG** [comment]
@LorenzMeier Hi, I rebased again and added the rc.mavlink_override script as discussed in the dev call.

Just as a side note, we use parameters to decide the mac address and IP address. In the past, it didn't seem to get too much appreciation, but I would gladly pr it again now if there's any interest.

**Comment by mrpollo** [comment]
@davids5 can you make sure that this falls into your plans for mavlink/ethernet networking support, and its aligned?

**Comment by davids5** [comment]
> @davids5 can you make sure that this falls into your plans for mavlink/ethernet networking support, and its aligned?

This is just the configuration and is not setting the nodes's IP address. So good to go from my perspective.

@garfieldG - Would you please rebase this, so we can test it and merge it.

**Comment by garfieldG** [comment]
@davids5 rebased it, thanks!

**Comment by davids5** [comment]
@garfieldG - Thank you, I have set it to auto merge. But if CI fails ping me.

**Comment by garfieldG** [comment]
@davids5 The CI failed

**Comment by mrpollo** [comment]
I suspect the failures have nothing to do with this PR, I restarted the jobs just to be sure, but the errors were mostly segmentation faults and timeouts.

I will monitor the results of CI and merge if they are indeed false positives.

File: Tools/serial/generate_config.py [hunk 2]
```diff
@@ -211,6 +224,80 @@ def parse_yaml_parameters_config(yaml_config):
         param_group = parameters_section.get('group', None)
         for param_name in definitions:
             param = definitions[param_name]
+            if 'ethernet' in param:
```
Comment (bkueng) [inline Tools/serial/generate_config.py:227]: Can you avoid the code duplication here, and instead pass `has_ethernet_support`?

File: Tools/serial/generate_config.py [hunk 3]
```diff
@@ -339,7 +428,8 @@ def parse_yaml_parameters_config(yaml_config):
             'port_param_name': port_param_name,
             'default_port': default_port,
             'param_group': port_config['group'],
-            'description_extended': port_config.get('description_extended', '')
+            'description_extended': port_config.get('description_extended', ''),
+			'ethernet_config': serial_command.get('ethernet', 'none')
```
Comment (bkueng) [inline Tools/serial/generate_config.py:432]: `'ethernet_config': serial_command.get('ethernet', None)`

File: validation/module_schema.yaml [hunk 4]
```diff
@@ -206,5 +210,9 @@ parameters:
                             # [0, N-1]
                             # Default: 0
                             type: integer
+                        ethernet:
```
Comment (bkueng) [inline validation/module_schema.yaml:213]: Can you change this to `requires_ethernet: type: boolean`?

**Pull request merged**